refactor(semantic): remove redundant code handling reference flags

Dunqing opened the pull request:
This PR is a follow-up to #11990. After `AstKind::TSTypeName` has been removed, a lot of prior handling of the kind can be removed. The semantic snapshot changes are expected, caused by #12042.

camchenry reviewed (approved):
Nice, looks like this might even be a +1-2% perf win too

<img width="749" alt="Image" src="https://github.com/user-attachments/assets/308b45dc-2f5b-4339-a345-26fb86a7d183" />

Dunqing committed:
refactor(semantic): remove redundant code handling reference flags (#12043)

This PR is a follow-up to #11990. After `AstKind::TSTypeName` has been removed, a lot of prior handling of the kind can be removed. The semantic snapshot changes are expected, caused by #12042.
